### ISSUE TITLE
Java architecture: enforce the semantic-volume and Voxy boundaries with ArchUnit

### ISSUE THREAD

**Issue by rhythmatician** [rhythmatician/voxygen-monorepo#184]
## Goal

Add project-owned architecture tests that make the already-settled Voxygen semantic and storage boundaries executable.

A normal Java unit-test run must fail when production code introduces a forbidden dependency across those boundaries.

This is a bounded architecture tracer. It protects three existing seams without attempting to modularize the entire Java source tree.

## Decision

The architecture is already settled:

- `VoxelVolume` is an opaque semantic XYZ volume of canonical block and biome IDs.
- `VoxelVolumeWriter` is the deep seam between generation and storage.
- Voxy packing, YZX indexing, reflection, `WorldSection`, and lifecycle mechanics remain behind the storage adapter.
- `VoxelPredictionDecoder` owns model-output layout and conversion into `VoxelVolume`; writers do not consume logits or raw model output.

Use ArchUnit as a normal JUnit-compatible test dependency. Pin `com.tngtech.archunit:archunit:1.5.0`. ArchUnit can test class and package dependencies, layers, slices, and cycles from compiled bytecode.

Most relevant Voxygen classes currently share the broad `com.rhythmatician.lodiffusion.voxy` package. Therefore:

- express the initial rules with explicit classes and class-family predicates;
- do not pretend the current package is one coherent architectural layer;
- do not move production classes merely to make a package-based ArchUnit diagram look cleaner;
- leave package extraction for a later issue if these rules demonstrate that it would improve the structure.

`CONTEXT.md` and the conforming definitions in `GLOSSARY.md` are the normative architecture inputs. Graphify may help locate dependencies, but it is not the enforcement oracle and its historical report must not be converted into a generated allowlist. The existing report is also a point-in-time graph rather than current architectural authority.

`rhythmatician/mattpocock-skills#4` is a non-blocking design reference, not an implementation dependency.

## Scope

Add `ArchitectureGuardrailsTest` under the Java unit-test source set. It must import production classes and enforce these rules.

### Rule 1: semantic core remains runtime-independent

The following semantic classes form the initial protected set:

- `SectionPos`
- `Level`
- `CanonicalRegistries`
- `VoxelVolume`
- `VoxelVolumeWriter`
- `WriteOutcome`
- `VolumeUnavailableException`
- `InMemoryVolumeWriter`

These classes must not depend on:

- `net.minecraft..`
- `net.fabricmc..`
- `me.cortex.voxy..`
- `ai.djl..`
- `java.lang.reflect..`
- `java.lang.invoke..`

Ordinary JDK collections, primitives, records, exceptions, and other semantic classes in the protected set remain legal.

This rule makes the current intended property of `InMemoryVolumeWriter` executable: it is deterministic and uses semantic values without Minecraft or Voxy runtime classes.

### Rule 2: writers do not understand model output

Every production implementation of `VoxelVolumeWriter` must be independent of:

- `com.rhythmatician.lodiffusion.onnx..`
- `ai.djl..`

A writer may receive only the semantic values exposed by the writer contract. It may not consume logits, tensors, `InferenceResult`, model-runner objects, or model-output array layouts.

This protects the existing contract that `VoxelPredictionDecoder` owns model-output layout while `VoxelVolumeWriter` exposes only semantic operations.

### Rule 3: the decoder does not reach into storage

`VoxelPredictionDecoder` must not depend on:

- `RealVoxyVolumeWriter`
- `VoxyCompat`
- `VoxyEngine`
- `VoxyWorldBinding`
- `VoxyProcessingAPI`
- `VoxyDetection`
- `VoxyBlockMapper`
- `VoxyIdMaps`
- `CanonicalVoxyMaps`
- `VoxySectionWriter`
- any `me.cortex.voxy..` type

The decoder may depend on semantic classes and deterministic fallback logic, but not on Voxy storage, reflection bindings, or writer implementations.

### Rule-sensitivity proof

Do not merely show that the current code passes.

Add test-only good and bad fixtures and prove:

1. a legal semantic dependency passes the corresponding rule;
2. a representative semantic class with a forbidden runtime dependency is rejected;
3. a representative `VoxelVolumeWriter` implementation with a model-runtime dependency is rejected.

The fixtures must live in test sources. Do not commit a deliberately invalid production class.

The rule construction used for the fixtures must be the same rule construction applied to production classes, rather than a separate imitation of it.

## Acceptance criteria

- [ ] ArchUnit `1.5.0` is pinned as a test dependency.
- [ ] `ArchitectureGuardrailsTest` runs through the existing `test` task.
- [ ] The semantic-core runtime-independence rule is enforced.
- [ ] The writer/model-output separation rule is enforced.
- [ ] The decoder/storage separation rule is enforced.
- [ ] Test-only fixtures prove at least one allowed dependency passes and representative forbidden dependencies fail.
- [ ] The architecture tests inspect production bytecode and exclude tests, generated classes, external references, and DataHarvester production code from the Voxygen rule population.
- [ ] No production package move, pass-through interface, or unrelated refactor is introduced solely to satisfy the rules.
- [ ] Existing Java behavior remains unchanged.
- [ ] The full current Java CI command succeeds.

## Verification

Run:

```text
./java/gradlew -p java test -PexcludeVoxyTestRuntime --tests '*ArchitectureGuardrailsTest'
./java/gradlew -p java lint compileJava compileClientJava
./java/gradlew -p java test -PexcludeVoxyTestRuntime
npm run typecheck
npm test
```

Demonstrate the sensitivity test in the test output. Do not temporarily modify production source as the proof.

## Dependencies

No native blockers.

`rhythmatician/mattpocock-skills#4` is reference material only.

## One-session sizing and vertical shape

This issue adds one dependency, one architecture-test surface, three concrete rules, and rule-sensitivity fixtures. It is sized for one fresh implementation session.

It crosses architecture definition, executable enforcement, the existing Gradle test task, and the current Java CI lane. No separate architecture service or agent is required.

## Out of scope

- Reorganizing the complete Java package tree.
- Refactoring `GenerationSession`, `LodGenerationService`, or unrelated god nodes.
- Enforcing subjective metrics such as “clean,” “simple,” or maximum class size.
- Converting the entire Graphify graph into an allowlist.
- Cross-language Python/Java dependency rules.
- Adding another GitHub Actions job or Merge Oracle.
- Modifying `CONTEXT.md`, `GLOSSARY.md`, or `AGENTS.md`.
- Automatically invoking `mattpocock-skills#4`.

**Comment by rhythmatician** [comment]
Sandcastle claiming #184 for AFK implementation on `sandcastle/issue-184` -- `Java architecture: enforce the semantic-volume and Voxy boundaries with ArchUnit`

**Comment by rhythmatician** [comment]
Sandcastle review rejected `sandcastle/issue-184` - not merged. Preserved branch for inspection.

**Verdict: approved=false**
**Summary:** Issue 184's ArchUnit implementation (commit 1bfd5dd) meets the contract: pinned 1.5.0, three rules with exact class/package sets, correct bytecode import exclusions, and same-construction sensitivity fixtures. Branch as staged is blocked because it bundles unrelated #152 production changes (PerformanceMonitor/GenerationSession tracer completion) violating 'no unrelated refactor' and 'behavior unchanged', and local Java 21 prevents verifying the required Java 25 CI command.

**Findings:**
- [blocking] Branch main...sandcastle/issue-184 bundles out-of-scope production behavior from commit 66f4a70 (#152): PerformanceMonitor.java TRACER_HORIZON_* constants + setCounter and GenerationSession.java TracerCompletion record, tracerWritten/Skipped/Failed counters, maybeEmitTracerTerminal, enqueue/processTracerRequest changes (160 lines) plus EndL4TracerCompletionTest.java. This violates AC 'No production package move, pass-through interface, or unrelated refactor is introduced solely to satisfy the rules' when evaluating the branch as staged.
- [blocking] AC 'Existing Java behavior remains unchanged' violated on branch main...HEAD: GenerationSession tracer changes alter runtime semantics (new terminal 121/121 logging, PerformanceMonitor coupling, stopRequested handling). In isolatio
**Acceptance criteria:**
- [x] ArchUnit `1.5.0` is pinned as a test dependency. - java/build.gradle:72 testImplementation 'com.tngtech.archunit:archunit:1.5.0' in commit 1bfd5dd
- [x] `ArchitectureGuardrailsTest` runs through the existing `test` task. - java/src/test/java/com/rhythmatician/lodiffusion/voxy/ArchitectureGuardrailsTest.java:21 JUnit 5 class under src/test, imported via ClassFileImporter; runs via standard Gradle test task (local run blocked by Java 25 requirement, not by test structure)
- [x] The semantic-core runtime-independence rule is enforced. - ArchitectureGuardrailsTest.java:23-41 SEMANTIC_CORE 8 classes, SEMANTIC_FORBIDDEN_PACKAGES 6 prefixes, semanticCoreRule() via onlyDependOnClassesThat(not in forbidden) and test semanticCore_isRuntimeIndependent
- [x] The writer/model-output separation ru
**Full verdict:** ```json
{
  "approved": false,
  "findings": [
    {
      "message": "Branch main...sandcastle/issue-184 bundles out-of-scope production behavior from commit 66f4a70 (#152): PerformanceMonitor.java TRACER_HORIZON_* constants + setCounter and GenerationSession.java TracerCompletion record, tracerWritten/Skipped/Failed counters, maybeEmitTracerTerminal, enqueue/processTracerRequest changes (160 lines) plus EndL4TracerCompletionTest.java. This violates AC 'No production package move, pass-through interface, or unrelated refactor is introduced solely to satisfy the rules' when evaluating the branch as staged.",
      "severity": "blocking"
    },
    {
      "message": "AC 'Existing Java behavior remains unchanged' violated on branch main...HEAD: GenerationSession tracer changes alter runtime semantics (new terminal 121/121 logging, PerformanceMonitor coupling, stopRequested handling). In isolation commit 1bfd5dd leaves behavior unchanged; the violation is due to branch history bundling.",
      "severity": "blocking"
    },
    {
      "message": "AC 'The full current Java CI command succeeds' not verifiable locally: ./java/gradlew test requires Java 25 (build.gradle release 25, CI setup-java 25) but runner has Temurin 21, build fails with 'release version 25 not supported'. CI success for ArchitectureGuardrailsTest must be verified on Java 25 runner.",
      "severity": "blocking"
    }
  ],
  "acceptanceCriteriaMet": [
    {
      "criterion": "ArchUnit `1.5.0` is pinned as a test dependency.",
      "met": true,
      "evidence": "java/build.gradle:72 testImplementation 'com.tngtech.archunit:archunit:1.5.0' in commit 1bfd5dd"
    },
    {
      "criterion": "`ArchitectureGuardrailsTest` runs through the existing `test` task.",
      "met": true,
      "evidence": "java/src/test/java/com/rhythmatician/lodiffusion/voxy/ArchitectureGuardrailsTest.java:21 JUnit 5 class under src/test, imported via ClassFileImporter; runs via standard Gradle test task (local run blocked by Java
```

Branch: `sandcastle/issue-184`

To retry: fix implementation to address findings, ensure `agent:blocked` is removed, `agent:implement` remains, and re-run factory.
